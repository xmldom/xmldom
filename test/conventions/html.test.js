'use strict';

const {
	HTML_BOOLEAN_ATTRIBUTES,
	isHTMLBooleanAttribute,
	HTML_RAW_TEXT_ELEMENTS,
	isHTMLRawTextElement,
	isHTMLEscapableRawTextElement,
	HTML_VOID_ELEMENTS,
	isHTMLVoidElement,
} = require('../../lib/conventions');

describe('HTML_BOOLEAN_ATTRIBUTES', () => {
	Object.keys(HTML_BOOLEAN_ATTRIBUTES).forEach((key) => {
		const value = HTML_BOOLEAN_ATTRIBUTES[key];
		it(`should contain immutable ${key} with value 'true'`, () => {
			expect([key, value]).toMatchSnapshot();
			try {
				HTML_BOOLEAN_ATTRIBUTES[key] = 'boo';
			} catch {}
			expect(HTML_BOOLEAN_ATTRIBUTES[key]).toBe(value);
		});
	});
});
describe('isHTMLBooleanAttribute', () => {
	Object.keys(HTML_BOOLEAN_ATTRIBUTES).forEach((key) => {
		it(`should detect attribute '${key}'`, () => {
			expect(isHTMLBooleanAttribute(key)).toBe(true);
		});
		const upperKey = key.toUpperCase();
		it(`should detect attribute '${upperKey}'`, () => {
			expect(isHTMLBooleanAttribute(upperKey)).toBe(true);
		});
		const mixedKey = key[0].toUpperCase() + key.substring(1);
		it(`should detect attribute '${mixedKey}'`, () => {
			expect(isHTMLBooleanAttribute(mixedKey)).toBe(true);
		});
	});
	it('should not detect prototype properties', () => {
		expect(isHTMLBooleanAttribute('hasOwnProperty')).toBe(false);
		expect(isHTMLBooleanAttribute('constructor')).toBe(false);
		expect(isHTMLBooleanAttribute('prototype')).toBe(false);
		expect(isHTMLBooleanAttribute('__proto__')).toBe(false);
	});
});
describe('HTML_VOID_ELEMENTS', () => {
	Object.keys(HTML_VOID_ELEMENTS).forEach((key) => {
		const value = HTML_VOID_ELEMENTS[key];
		it(`should contain immutable ${key} with value 'true'`, () => {
			expect([key, value]).toMatchSnapshot();
			try {
				HTML_VOID_ELEMENTS[key] = 'boo';
			} catch {}
			expect(HTML_VOID_ELEMENTS[key]).toBe(true);
		});
	});
});
describe('isHTMLVoidElement', () => {
	Object.keys(HTML_VOID_ELEMENTS).forEach((key) => {
		it(`should detect attribute '${key}'`, () => {
			expect(isHTMLVoidElement(key)).toBe(true);
		});
		const upperKey = key.toUpperCase();
		it(`should detect attribute '${upperKey}'`, () => {
			expect(isHTMLVoidElement(upperKey)).toBe(true);
		});
		const mixedKey = key[0].toUpperCase() + key.substring(1);
		it(`should detect attribute '${mixedKey}'`, () => {
			expect(isHTMLVoidElement(mixedKey)).toBe(true);
		});
	});
	it('should not detect prototype properties', () => {
		expect(isHTMLVoidElement('hasOwnProperty')).toBe(false);
		expect(isHTMLVoidElement('constructor')).toBe(false);
		expect(isHTMLVoidElement('prototype')).toBe(false);
		expect(isHTMLVoidElement('__proto__')).toBe(false);
	});
});
describe('HTML_RAW_TEXT_ELEMENTS', () => {
	Object.keys(HTML_RAW_TEXT_ELEMENTS).forEach((key) => {
		const value = HTML_RAW_TEXT_ELEMENTS[key];
		it(`should contain immutable ${key} with value 'true'`, () => {
			expect([key, value]).toMatchSnapshot();
			try {
				HTML_RAW_TEXT_ELEMENTS[key] = 'boo';
			} catch {}
			expect(HTML_RAW_TEXT_ELEMENTS[key]).toBe(value);
		});
	});
});
describe('isHTMLRawTextElement', () => {
	Object.keys(HTML_RAW_TEXT_ELEMENTS).forEach((key) => {
		const expected = HTML_RAW_TEXT_ELEMENTS[key] === false;
		it(`should detect attribute '${key}' as ${expected}`, () => {
			expect(isHTMLRawTextElement(key)).toBe(expected);
		});
		const upperKey = key.toUpperCase();
		it(`should detect attribute '${upperKey}' as ${expected}`, () => {
			expect(isHTMLRawTextElement(upperKey)).toBe(expected);
		});
		const mixedKey = key[0].toUpperCase() + key.substring(1);
		it(`should detect attribute '${mixedKey}' as ${expected}`, () => {
			expect(isHTMLRawTextElement(mixedKey)).toBe(expected);
		});
	});
	it('should not detect prototype properties', () => {
		expect(isHTMLRawTextElement('hasOwnProperty')).toBe(false);
		expect(isHTMLRawTextElement('constructor')).toBe(false);
		expect(isHTMLRawTextElement('prototype')).toBe(false);
		expect(isHTMLRawTextElement('__proto__')).toBe(false);
	});
});
describe('isHTMLEscapableRawTextElement', () => {
	Object.keys(HTML_RAW_TEXT_ELEMENTS).forEach((key) => {
		const expected = HTML_RAW_TEXT_ELEMENTS[key];
		it(`should detect attribute '${key}' as ${expected}`, () => {
			expect(isHTMLEscapableRawTextElement(key)).toBe(expected);
		});
		const upperKey = key.toUpperCase();
		it(`should detect attribute '${upperKey}' as ${expected}`, () => {
			expect(isHTMLEscapableRawTextElement(upperKey)).toBe(expected);
		});
		const mixedKey = key[0].toUpperCase() + key.substring(1);
		it(`should detect attribute '${mixedKey}' as ${expected}`, () => {
			expect(isHTMLEscapableRawTextElement(mixedKey)).toBe(expected);
		});
	});
	it('should not detect prototype properties', () => {
		expect(isHTMLEscapableRawTextElement('hasOwnProperty')).toBe(false);
		expect(isHTMLEscapableRawTextElement('constructor')).toBe(false);
		expect(isHTMLEscapableRawTextElement('prototype')).toBe(false);
		expect(isHTMLEscapableRawTextElement('__proto__')).toBe(false);
	});
});
