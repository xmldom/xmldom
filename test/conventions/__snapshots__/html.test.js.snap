// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable allowfullscreen with value 'true' 1`] = `
Array [
  "allowfullscreen",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable async with value 'true' 1`] = `
Array [
  "async",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable autofocus with value 'true' 1`] = `
Array [
  "autofocus",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable autoplay with value 'true' 1`] = `
Array [
  "autoplay",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable checked with value 'true' 1`] = `
Array [
  "checked",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable controls with value 'true' 1`] = `
Array [
  "controls",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable default with value 'true' 1`] = `
Array [
  "default",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable defer with value 'true' 1`] = `
Array [
  "defer",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable disabled with value 'true' 1`] = `
Array [
  "disabled",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable formnovalidate with value 'true' 1`] = `
Array [
  "formnovalidate",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable hidden with value 'true' 1`] = `
Array [
  "hidden",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable ismap with value 'true' 1`] = `
Array [
  "ismap",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable itemscope with value 'true' 1`] = `
Array [
  "itemscope",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable loop with value 'true' 1`] = `
Array [
  "loop",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable multiple with value 'true' 1`] = `
Array [
  "multiple",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable muted with value 'true' 1`] = `
Array [
  "muted",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable nomodule with value 'true' 1`] = `
Array [
  "nomodule",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable novalidate with value 'true' 1`] = `
Array [
  "novalidate",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable open with value 'true' 1`] = `
Array [
  "open",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable playsinline with value 'true' 1`] = `
Array [
  "playsinline",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable readonly with value 'true' 1`] = `
Array [
  "readonly",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable required with value 'true' 1`] = `
Array [
  "required",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable reversed with value 'true' 1`] = `
Array [
  "reversed",
  true,
]
`;

exports[`HTML_BOOLEAN_ATTRIBUTES should contain immutable selected with value 'true' 1`] = `
Array [
  "selected",
  true,
]
`;

exports[`HTML_RAW_TEXT_ELEMENTS should contain immutable script with value 'true' 1`] = `
Array [
  "script",
  false,
]
`;

exports[`HTML_RAW_TEXT_ELEMENTS should contain immutable style with value 'true' 1`] = `
Array [
  "style",
  false,
]
`;

exports[`HTML_RAW_TEXT_ELEMENTS should contain immutable textarea with value 'true' 1`] = `
Array [
  "textarea",
  true,
]
`;

exports[`HTML_RAW_TEXT_ELEMENTS should contain immutable title with value 'true' 1`] = `
Array [
  "title",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable area with value 'true' 1`] = `
Array [
  "area",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable base with value 'true' 1`] = `
Array [
  "base",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable br with value 'true' 1`] = `
Array [
  "br",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable col with value 'true' 1`] = `
Array [
  "col",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable embed with value 'true' 1`] = `
Array [
  "embed",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable hr with value 'true' 1`] = `
Array [
  "hr",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable img with value 'true' 1`] = `
Array [
  "img",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable input with value 'true' 1`] = `
Array [
  "input",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable link with value 'true' 1`] = `
Array [
  "link",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable meta with value 'true' 1`] = `
Array [
  "meta",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable param with value 'true' 1`] = `
Array [
  "param",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable source with value 'true' 1`] = `
Array [
  "source",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable track with value 'true' 1`] = `
Array [
  "track",
  true,
]
`;

exports[`HTML_VOID_ELEMENTS should contain immutable wbr with value 'true' 1`] = `
Array [
  "wbr",
  true,
]
`;
